Standalone electron app

ptomato opened the pull request:
https://phabricator.endlessm.com/T20603

ptomato committed:
Change to standalone, and standardize "hatch-previewer" name

This makes the necessary changes for hatch-previewer to be its own npm
module, and standardizes on the name "hatch-previewer" everywhere.

See libingester#123.

https://phabricator.endlessm.com/T20603

sgnn7 committed:
Fixed CSS path loading

Depending on how we are running, we need to find our styles in the
node_modules directory which may be in different path. Loading these
sequentially generally doesn't cause problems as errors are silently
ignored and found CSS styling is applied.

ptomato committed:
Replace console.log with structured logging

We don't want debug output by default, so use a prominent Node.JS
logging package for log levels.

Get full debug output with DEBUG=hatch-previewer.

https://phabricator.endlessm.com/T20603

ptomato committed:
Add linting

ESLint is run on "npm test", and the configuration file is based on
eslint:recommended, eslint-config-node, and taking the file generated by
ESLint examining the files and tweaking it to fit the JS style we use in
other projects.

eslint-plugin-html is used to lint the inline <script> tags in the HTML.

Fixes any linter errors that eslint shows.

https://phabricator.endlessm.com/T20603

ptomato committed:
Fix invalid HTML comment

'--' is not allowed in an HTML comment unless it is immediately followed
by '>'.

https://phabricator.endlessm.com/T20603

ptomato committed:
Bump version

https://phabricator.endlessm.com/T20603

sgnn7 commented on bin/hatch-previewer:
```diff
@@ -0,0 +1,24 @@
+
+// Start app
+var child = proc.spawn(electron,
+                       ['.', ...passedArgs],
```
Have you tried testing this in all 3 envs?
- Within hatch-previewer
- Within l-i
- Within an ingester with previewer as a dependency

 I feel like this might not work in all of them.

ptomato replied:
My opinion is that we don't want hatch-previewer within libingester anymore, since part of the reason for this split is to lose the heavy electron dependency for every ingester. So I don't think the latter two are still use cases that need to be supported.

sgnn7 replied:
Even if you exclude that, let's say you have a dev dep on this lib in an ingester - can the new code work fine in that case when you invoke `npm previewer` or `node_modules/.bin/hatch-preview` with a relative path (I can't tell for sure but I'd say it's unlikely)?

The use case here is that you can both preview a hatch from this repo itself using relative/absolute paths but also preview hatches from one of the ingester directories using also relative or absolute paths invoking this script in the `node_modules/.bin/` directory.

ptomato replied:
I don't understand the use case for having hatch-previewer as a dev dependency after this, since hatch-previewer and libingester aren't tied together. Wouldn't it be better to install it with `npm install -g` or as a flatpak?

ptomato replied:
On re-examining the code, I still don't see why this is needed. It looks up the path of the libingester module that hatch-previewer is part of, but this module isn't part of libingester anymore. If it's not part of libingester, then we don't need to change the CWD. If we don't change the CWD, then we don't need to resolve a relative path passed on the command line.

sgnn7 commented on main.js:
```diff
@@ -7,6 +7,20 @@ const BrowserWindow = electron.BrowserWindow;
+
+let level = 'warn'
+if (process.env.DEBUG && process.env.DEBUG.includes('hatch-previewer'))
+  level = 'debug'
```
4 spaces indent. might also be better with braces even if it's a single-line conditional

ptomato replied:
I was trying to stick to the style that was in the majority of JS code, but we can certainly fix it all up at once. See comment below.

sgnn7 commented on main.js:
```diff
@@ -7,6 +7,20 @@ const BrowserWindow = electron.BrowserWindow;
 const path = require('path');
 const url = require('url');
 const fs = require('fs');
+const winston = require('winston')
```
I know it's outside of the scope of conversion but I think we should just bite the bullet and semicolon fix all the js files in this pass. Thoughts on this?

ptomato replied:
Fine with me! Eslint should make that easy. I'm not clear on what the desired style is, though; do you want semicolons required or semicolons forbidden?

sgnn7 replied:
Required semicolons for sure :)

ptomato committed:
Fix size of app icon

The app icon was 256x257, with an extra blank row of pixels at the top.
This fixes it to be the required size of 256x256. Also moves it to an
icons/ directory and adjusts the location where the app looks for it,
and renames it to the app ID which will be required for Flathub.

https://phabricator.endlessm.com/T20603

sgnn7 reviewed (changes requested):
@ptomato Just a couple of comments

ptomato committed:
Add the trappings of a Flathub app

Flathub requires an appdata file, a desktop file, icons at 128 and 64
pixels, and screenshots.

https://phabricator.endlessm.com/T20603

sgnn7 commented:
@ptomato re relative path loading in the launcher script: let's just leave it nixed as you have it and we can see if any use cases break. it should be relatively easy to add back if something isn't working but it's probably not worth it to spend a lot of time on blocking this PR for so skip it for now.

ptomato committed:
main: Open file dialog if no hatch folder given

Previously the app would exit if no hatch folder was given on the
command line. This instead opens a file dialog to let the user choose a
hatch folder if not passing one on the command line.

https://phabricator.endlessm.com/T20603

dnohales commented on com.endlessm.HatchPreviewer.appdata.xml:
```diff
@@ -0,0 +1,64 @@
+    <content_attribute id="money-gambling">none</content_attribute>
+  </content_rating>
+  <update_contact>philip_AT_endlessm.com</update_contact>
+</component>
```
No blank line at the end of file.

dnohales commented on com.endlessm.HatchPreviewer.desktop:
```diff
@@ -0,0 +1,6 @@
+Type=Application
+Name=Hatch Previewer
+Icon=com.endlessm.HatchPreviewer
+Exec=/app/bin/run.sh %f
```
Is it not supposed to be `/app/bin/hatch-previewer %f`?

sgnn7 replied:
I'd +1 this

ptomato replied:
Oh, darn. The build stuff for flatpak requires a wrapper script here. But I guess if I call the wrapper script `hatch-previewer`, then the invocations can be the same 😝

sgnn7 commented on README.md:
```diff
@@ -0,0 +1,15 @@
+### Usage ###
+
+- From this repo
+ - `npm start -- <path>`
```
this should probably use the runner script (`./bin/hatch-preview`) instead of npm start

sgnn7 commented on main.js:
```diff
@@ -1,115 +1,120 @@
+
+let level = 'warn';
+if (process.env.DEBUG && process.env.DEBUG.includes('hatch-previewer'))
+    level = 'debug';
```
Maybe wrap the case in parens or space the next line one more emty newline down? It's a bit confusing to read given it's visual context.

sgnn7 commented on preview_renderer.js:
```diff
@@ -4,91 +4,97 @@ const $ = require('jquery');
-    $('#preview_frame').html("<center><h2>No asset selected</h2></center>")
-    return;
-  }
+window.flipPage = function () {
```
Do we need all of these internal methods attached to the window? I'm seeing this in other files too.

ptomato replied:
These are used outside of the file where they're defined, so I'm not sure where else to put them. Maybe `exports` would work?

ptomato replied:
At least `preview` and `flipPage` have to be defined on `window`, since they're used from `onclick` handlers in the HTML.

ptomato replied:
I fixed the other ones to use `exports` and have the modules where they're used `require` them directly.

sgnn7 replied:
:+1:

sgnn7 reviewed (changes requested):
@ptomato Couple of small comments

sgnn7 committed:
Made hatch location resolution use absolute paths

Old code was not using proper paths when invoked as a dependency module
due to relative path differences so now we make sure to use an absolute
path.

sgnn7 commented:
CR Pass!
